Added support for LittlevGL library

vanwinkeljan opened the pull request:
Added support for LittlevGL GUI library.

From [LittlevGL](https://littlevgl.com/) website
> LittlevGL is a free and open-source graphics library providing everything you need to create embedded GUI with easy-to-use graphical elements, beautiful visual effects and low memory footprint.

* Requested by #6709
* Requested by #2748

vanwinkeljan committed:
gui: Added glue logic for LittlevGL GUI library

Added glue logic to interface Zephyr with LittlevGL GUI library

This includes:
 * KConfig options for all lvgl options
 * Kernel & user space memory management
 * Zephyr to lvgl FS call mapping
 * Color space conversion function

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

vanwinkeljan committed:
sample: gui: Added basic sample for lvgl library

Added a basic sample showing how to use the LittlevGL library in an
application.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

carlescufi commented:
@vanwinkeljan can you rebase this against the latest master? There's a bunch of commits that don't belong to the PR

vanwinkeljan commented:
@carlescufi What is the normal approach if you have dependencies to other PR?
I could issue a PR without the dependencies but then nobody can test it.

lpereira commented:
While I can imagine this being useful, this adds a lot of code (over 180k lines) that has to be tracked and updated each Zephyr release.

nashif commented:
> While I can imagine this being useful, this adds a lot of code (over 180k lines) that has to be tracked and updated each Zephyr release.

this will hopefully be one of those dependencies that will have to be maintained externally to zephyr, i.e. in its own repo in the project or similar to openthread.

nashif commented:
> @vanwinkeljan can you rebase this against the latest master? There's a bunch of commits that don't belong to the PR

those are not-merged patches from the SPI PR..

carlescufi commented:
>those are not-merged patches from the SPI PR..

@vanwinkeljan @nashif sorry, I didn't realize. Sorry for the noise.

vanwinkeljan commented:
@nashif @laperie I had a look at how this is handled for openthread, as lvgl has no dedicated makefiles I used the FetchContent module instead. Only downside is that the FetchContent module was only introduced in cmake 3.11

jukkar commented:
Thanks for pushing this. It would be great to get some gui support for zephyr. I am also hoping that someone would create a driver for the small oled displays that are used a lot in this field.

> I had a look at how this is handled for openthread, as lvgl has no dedicated makefiles I used the FetchContent module instead.

I did not quite understand your comment about makefiles. Could you elaborate a bit more why we would not do similar things as what subsys/net/lib/openthread is doing? Adding a version requirement for cmake should be avoided IMHO.

vanwinkeljan commented:
@jukkar Problem is that [lvgl repo](https://github.com/littlevgl/lvgl) has no top level makefile so if we would like to use the [ExternallModule](https://cmake.org/cmake/help/latest/module/ExternalProject.html) cmake module instead of a [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html) module a top level makefile needs to be added after fetching the lvgl repo, in case of the FetchContent module ext/lib/gui/lvgl/CMakeLists.txt can be used to build the lvgl library.

Further note that there is already a requirement for a minimal cmake version (3.8.2) and I just bumped it to 3.11.

jfischer-phytec-iot commented on drivers/display/display_ili9340.c:
```diff
@@ -0,0 +1,184 @@
+	ili9340_write_bitmap(dev, x, y, 1, 1, &rgb_data[0]);
+}
+
+void ili9340_write_bitmap(const struct device *dev, const u16_t x,
```
why not ili9340_write_ram or ili9340_write_display_ram ?

vanwinkeljan replied:
As this is an external interface and it is actually writing a bitmap (rectangular image) and not a flat/consecutive data array

jfischer-phytec-iot commented on include/display.h:
```diff
@@ -0,0 +1,117 @@
+ */
+typedef int (*display_write_bitmap_api)(const struct device *dev, const u16_t x,
+					const u16_t y, const u16_t w,
+					const u16_t h, const u8_t *rgb_data);
```
rgb_data is confusing because the display must not be a RGB display, just data is sufficient.

vanwinkeljan replied:
Good point will take this change along,

jfischer-phytec-iot commented on include/display.h:
```diff
@@ -0,0 +1,117 @@
+ * @param y y coordinate of the upper left corner
+ * @param w width of the bitmap
+ * @param h height of the bitmap
+ * @param rgb_data pointer to the RGB data array, the data array should be at
```
the same here, is it a API for RGB display drivers?

dbkinder reviewed (approved):
OK on the docs, thanks!

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -27,13 +27,11 @@ if(uname_output MATCHES "MSYS")
+# set to OLD to avoid copy-pasting cmake_minimum_required across application
 # CMakeLists.txt files.
-cmake_minimum_required(VERSION 3.8.2)
+cmake_minimum_required(VERSION 3.11)
```
3.11 is too recent unfortunately.

The most recent Ubuntu LTS, 18.04, has 3.10.2 in it's package repository.

Using a CMake version that is not easily available makes it more difficult to get started with Zephyr.

FetchModule can be emulated through recursively invoking CMake, not pretty, but it
is the better solution for our users.

vanwinkeljan replied:
Actually this is a part I always have been surprised about. The zephyr documentation mentions the use of pip and on the other hand manual installation of cmake while the latest cmake version is available through pip (https://pypi.org/project/cmake/). 
E.g. the only thing I had to do was "pip install --upgrade cmake" and I had 3.11 available on Ubuntu LTS 16.04.

SebastianBoe replied:
I was not aware that one could install non-python packages through pip.

It's not obvious to me that we would want to encourage this in general.
Generally speaking users (rightly IMHO) prefer the distribution
package managers over pip.

But perhaps for CMake this makes sense ...

In any case using pip would require a separate PR.

andrewboie replied:
I have no problem with adding cmake to our pip requirements.txt, we already require a ton of stuff in it.

carlescufi replied:
We need to check whether this would work on Windows, and the documentation should be updated. I am somewhat skeptical personally, especially given that, if users already have a copy of CMake on their system, this might install a second one.

carlescufi replied:
That said I am open to this if it really means an improvement for our users and works reliably on Linux, macOS and Windows. It's an interesting discovery actually @vanwinkeljan, I had no idea one could do this and it might end up being a good thing indeed.

vanwinkeljan replied:
A general remark here is that for python it is best practice to install project dependencies in a dedicated virtual environment and not system wide (either through pip or a packet manager), in this way the system does not get polluted with unnecessary installs. Now the nice thing is if you install cmake through pip into a dedicated virtual environment it can coexist with a already system wide installed version.

vanwinkeljan commented:
@nashif could you label this pull request area: Display it is dependency of #9149?

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+#define LV_MEM_AUTO_DEFRAG  1               /*Automatically defrag on free*/
+#else       /*LV_MEM_CUSTOM*/
+#define LV_MEM_CUSTOM_INCLUDE <kernel.h>   /*Header for the dynamic memory function*/
+#define LV_MEM_CUSTOM_ALLOC   k_malloc       /*Wrapper to malloc*/
```
Moved from #9149
> @pizi-nordic
> What about using this library from userspace (k_malloc() is not available there)?
> What about using dedicated memory pool for graphics (in order to prevent memory exhausting by non-critical subsystem)?

vanwinkeljan replied:
@pizi-nordic 
Added support for heap and memory pools either in kernel or user space

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+ *=================*/
+
+/*Input device settings*/
+#define LV_INDEV_READ_PERIOD            50                     /*Input device read period in milliseconds*/
```
Moved form #9149 
> @pizi-nordic
>Shouldn't we make this configurable?

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+
+/*Text settings*/
+#define LV_TXT_UTF8           CONFIG_LVGL_UTF_8  /*Enable UTF-8 coded Unicode character usage */
+#define LV_TXT_BREAK_CHARS    " ,.;:-_"          /*Can break texts on these chars*/
```
Moved form #9149 
> @pizi-nordic
>IMHO this one should be also configurable.

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+/*Compiler settings*/
+#define LV_ATTRIBUTE_TICK_INC                 /* Define a custom attribute to `lv_tick_inc` function */
+#define LV_ATTRIBUTE_TASK_HANDLER             /* Define a custom attribute to `lv_task_handler` function */
+#define LV_COMPILER_VLA_SUPPORTED    1        /* 1: Variable length array is supported*/
```
Moved from #9149
> @pizi-nordic
> VLA might internally either use malloc() (which is not available in kernel scope) or alloca() (which allocates memory on stack which is in most case as small as possible). IMHO we should allow user to customize this setting basing on his needs.

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+/* More info about fonts: https://littlevgl.com/basics#fonts
+ * To enable a built-in font use 1,2,4 or 8 values
+ * which will determine the bit-per-pixel */
+#define LV_FONT_DEFAULT        &lv_font_dejavu_20     /*Always set a default font from the built-in fonts*/
```
Moved from #9149 
> @pizi-nordic
> You are always selecting lv_font_dejavu_20, but it can be compiled out basing on CONFIG_LVGL_BUILD_IN_FONT_20. Is that correct?

vanwinkeljan replied:
made this configurable including a pointer to a custom font

vanwinkeljan commented on samples/gui/lvgl/src/main.c:
```diff
@@ -0,0 +1,104 @@
+	lv_flush_ready();
+}
+
+void lvgl_tick_handler(struct k_timer *timer_id)
```
Moved from #9149 
> @pizi-nordic
> Is this function generic? If so, could we move it to the lvgl integration layer?

vanwinkeljan replied:
Yes it is and it only used to keep track of the system time in lvgl. So I'm wondering if it is not better to reimplement the lvgl functions ([lv_hal_tick.c](https://github.com/littlevgl/lvgl/blob/fc05fad2770d62de8085c88e1a2b15c313576909/lv_hal/lv_hal_tick.c)) to get the system time and time delta by a zephyr specific implementation.

vanwinkeljan replied:
lvgl v5.2-rc provides option to use zephyr system time, code has been update to make use of this feature

pizi-nordic commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,350 @@
+ * Required for buffered drawing, opacity and anti-aliasing
+ * VDB makes the double buffering, you don't need to deal with it!
+ * Typical size: ~1/10 screen */
+#define LV_VDB_SIZE         CONFIG_LVGL_VDB_SIZE  /*Size of VDB in pixel count (1/10 screen size is good for first)*/
```
Could we define this as percentage of display?
```
#define LV_VDB_SIZE (CONFIG_LVGL_VDB_SIZE  * CONFIG_LVGL_HOR_RES * CONFIG_LVGL_VER_RES / 100)
```
 IMHO it will be easier to use.

vanwinkeljan replied:
good suggestion, I will update the code

pizi-nordic commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,350 @@
+#else
+#error No color depth defined
+#endif
+#define LV_COLOR_TRANSP    LV_COLOR_LIME          /*Images pixels with this color will not be drawn (with chroma keying)*/
```
Shouldn't we allow for changing this value through Kconfig?

arturo182 commented on ext/lib/gui/lvgl/Kconfig:
```diff
@@ -0,0 +1,292 @@
+    config LVGL_COLOR_DEPTH_24
+        bool "24-bit"
+
+    config LVGL_COLOR_DEPTH_8
```
This should be `LVGL_COLOR_DEPTH_16`

arturo182 commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,363 @@
+/*================
+ *  THEME USAGE
+ *================*/
+#ifdef CONFIG_LGVL_THEMES
```
This should be `CONFIG_LVGL_THEMES` (LVGL is misspelled)

vanwinkeljan commented:
Updated PR to use LittlevGL version v5.2-rc
 For a list of changes see [LittlevGL release notes v5.2-rc](https://github.com/littlevgl/lvgl/releases/tag/v5.2-rc)

Update in this push:
* Update lv_conf.h to include new features
* Integrated lvgl logging with zephyr logging framework
* Added zephyr system init for lvgl library
* Added support for conversion functions between pixel format used in lvgl and
format used by display driver
* Use west instead of cmake FetchContent, this removes dependency towards cmake 3.11

arturo182 commented:
This is great! Just yesterday I took this PR and updated the lib to 5.2-rc and added 16_SWAP support so I could push pixels directly from lvgvl to a ILI3941 set up to also use 16 bit colors. You can see it here https://twitter.com/arturo182/status/1054117243070636032

These changes are even better cause they remove boilerplate code from main.c!

One thing @vanwinkeljan, I think you must rebase this PR against master so only relevant commits are included.

vanwinkeljan commented:
> This is great! Just yesterday I took this PR and updated the lib to 5.2-rc and added 16_SWAP support so I could push pixels directly from lvgvl to a ILI3941 set up to also use 16 bit colors. You can see it here https://twitter.com/arturo182/status/1054117243070636032

My apologize it came late, make sure to clone lvgl next to your zephyr repo if you switch (or use west + the new manifest). Nice project btw!

> 
> These changes are even better cause they remove boilerplate code from main.c!
> 

Actually I was also think about adding a GUI thread that takes care  of periodically calling ```lv_task_handler();``` and a user provided callback function.

> One thing @vanwinkeljan, I think you must rebase this PR against master so only relevant commits are included.

Actually the order was changed between the two PRs (#9149 and this one) so that it can be merged without needing lvgl. I assume that PR #9149 should not give you any problems unless you try to build for the native_posix target but even then it should just work if you have SDL 2.0 installed.

arturo182 commented on ext/lib/gui/lvgl/Kconfig.fonts:
```diff
@@ -0,0 +1,189 @@
+    help
+      Bits per pixel (1,2,4,8) for font, 0 will disable font
+
+config LVGL_BUILD_IN_FONT_10_SYMBOLS_BPP
```
You use `_SYMBOLS_` here...

arturo182 commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,464 @@
+#define USE_LV_FONT_DEJAVU_10              CONFIG_LVGL_BUILD_IN_FONT_10_ASCII_BPP
+#define USE_LV_FONT_DEJAVU_10_LATIN_SUP    CONFIG_LVGL_BUILD_IN_FONT_10_LATIN_SUP_BPP
+#define USE_LV_FONT_DEJAVU_10_CYRILLIC     CONFIG_LVGL_BUILD_IN_FONT_10_CYRILLIC_BPP
+#define USE_LV_FONT_SYMBOL_10              CONFIG_LVGL_BUILD_IN_FONT_10_SYMBOL_BPP
```
But `_SYMBOL_` here (note the missing "S"). The same is done for the other fonts.

arturo182 commented:
I think it would be interesting to look into implementing `lv_fs_drv_t` using Zephyr's file io at some point :D

arturo182 commented on ext/lib/gui/lvgl/Kconfig:
```diff
@@ -0,0 +1,345 @@
+if LVGL_MEM_POOL_KERNEL || LVGL_MEM_POOL_USER
+
+config LVGL_MEM_POOL_MIN_SIZE
+   int "Minimum memory pool bock size"
```
block is misspelled

arturo182 commented on ext/lib/gui/lvgl/Kconfig:
```diff
@@ -0,0 +1,345 @@
+     Size of the smallest block in the memory pool in bytes
+
+config LVGL_MEM_POOL_MAX_SIZE
+   int "Maximum memory pool bock size"
```
also here

vanwinkeljan commented:
@arturo182 FS support added

arturo182 commented:
Amazing! Can't wait to test it out :)

hongshui3000 commented:
I want to use sd card to store my picture, but I found that zephyr does not provide a driver :）

vanwinkeljan commented:
Added support for monochrome displays, tested on reel board.

![image](https://user-images.githubusercontent.com/1586475/48445211-4bf10f00-e796-11e8-9d4e-a7c1a134961a.png)

vanwinkeljan committed:
display: Added double buffer screen info to display API

Extend display API screen info with double buffer enum entry.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

MaureenHelm commented:
I got this working on the mimxrt1050_evk board

![image](https://user-images.githubusercontent.com/6172539/50645883-43b58480-0f3a-11e9-92a6-092a96b67860.png)

vanwinkeljan commented:
Updates:
* Moved glue logic code from ext/lib/gui/lvgl to lib/gui/lvgl
* Included sources of lvgl V5.2 in ext/lib/gui/lvgl

vanwinkeljan commented:
@SebastianBoe revisit please

nashif commented on include/gui/lvgl_mem.h:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2018 Jan Van Winkel <jan.van_winkel@dxplore.eu>
```
is this a public API of Zephyr? We should not have things under include/ that are not public APIs. We are in the process of cleaning up and we should not add new non-public headers in here.

vanwinkeljan replied:
No it is not a public API, moved header file to lib/gui/lvgl

pfalcon reviewed (dismissed):
Why do we add gazillion of files not directly related to being "The Zephyr Project is a scalable real-time operating system (RTOS) supporting multiple hardware architectures, optimized for resource constrained devices, and built with security in mind." ? Beyond that, who's going to update this codedrop? Why endure commit content noise related to these updates? 

Why not git submodule? Why not modules supported by that west tool?

vanwinkeljan commented:
@pfalcon Adding LVGL directly in to ext is a temporary solution until multi repo support is available.

See slak for more info: https://zephyrproject.slack.com/archives/C18PLHN8H/p1546535159243100

embeddedt commented:
I am one of the maintainers on the LittlevGL project and I have been following this pull request. A possible solution would be to make a Zephyr "glue" driver and add it to LittlevGL, rather than adding LittlevGL to Zephyr.

vanwinkeljan commented:
Problem is that we always need something at the zephyr side to integrate with the zephyr build environment and actually this is the majority of the _glue_ code (KConfig files, CMakeLists.txt & lv_conf.h). 

Adding the full lvgl code base to zephyr is only a temporary solution until _west_ is available, which will take care of fetching fetch the sources of external dependencies. 

Until _west_ is available we have two options and option 1 was selected:

1. Copy all LVGL source to ext/lib/giu/lvgl
2. Use CMake FetchContent

embeddedt commented:
@vanwinkeljan Can I ask why you don't use submodules instead? It is cleaner than duplicating the codebase inside another repository.

vanwinkeljan commented:
@embeddedt Because git submodules where not selected to maintain external dependencies but it was opted to use a _repo_  like tool (_west_), unfortunately this is not yet available for prime time.

vanwinkeljan commented:
@nashif @pfalcon @embeddedt In the end I pushed the CMake ```FetchContent``` solution and removed the lvgl source from ext/lib

pfalcon commented:
@nashif: I also wanted to tell that for a long time, just couldn't think of a good way do that. So let me just do it straight: github does not notify in any way an author of a review when their review is dismissed. Consequently, if you want to give an author of such a review an opportunity to check that the issues were actually fixed, they need to be somehow notified - maybe by including @mention in the dismissal description, but I'm not sure if that works (need to test), and otherwise by a straight comment with mention. And conversely, if the aim is to sneak some code past the normal review process and leave the impression that all PRs are approved unanimously, make sure to not mention the dismissal to affected persons, and they'll never know ;-).

embeddedt commented:
@pfalcon 

> Well, we're already regularly late with their updates, delivering stale/buggy content the users.

As a heads-up: we're planning to release the next version of LittlevGL (5.3) this month. The release after that is planned for June.

pfalcon commented:
@embeddedt: Thanks for heads-up.

Generally, my idea is that people should be motivated to work directly with upstream projects (bug fixes, etc.). And including an upstream project by reference motivates doing that, whereas including it by copy arguably demotivates that, driving towards "compromises" like "oh, we need to cut the release soon, let's just push that fix to our copy, we'll deal with all that later". And later only becomes more and more later. And as pointed by @nashif, we already doing/have to do that with vendor HALs. Well, that should be the motivation to NOT do that with community-maintained open source projects.

nashif commented on samples/gui/lvgl/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  name: lvgl
+tests:
+  test:
+    build_only: true
```
remove build_only and add
```
harness: display
```

nashif commented on samples/gui/lvgl/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+tests:
+  test:
+    build_only: true
+    platform_whitelist: nrf52840_pca10056, native_posix, reel_board
```
remove the commas

vanwinkeljan commented:
@nashif Updated the sample.yaml file, I did remove _native POSIX_ target as it depends on SDL2. I don't think it is included in the CI docker container and as suche the build will fail.
We have three options here:
 1. Add SDL2 to the CI image
 2. Use the dummy display driver through a overlay conf file
 3. Do not include native posix target in CI

Another question should I add myself to CODEWONERS for lib/gui ?

kisvegabor commented:
Hi,

@vanwinkeljan [notified us](https://github.com/littlevgl/lvgl/issues/715) at about you have integrated LittlevGL. As the author of LittlevGL I was very happy to hear that! 

My question is if you would be interested in writing a blog post on [LittlevGL's blog](https://blog.littlevgl.com/) about getting started with Zephyr and LittlevGL. 

I found [this description](https://docs.zephyrproject.org/latest/samples/gui/lvgl/README.html#) but I meant a something that is for beginners with Zephyr. It could a good option to introduce and promote Zephyr.

So would you be interested?

marc-hb commented:
> West is not ready, this PR has been open for almost a year now waiting for west or something similar, it is time to move on and use west when west is supported and ready for this.

West is in full swing now. Is there already a github issue filed or work in progress to move LittleVGL to west? If not I can file one. I'd like it not to be low priority because right now this component seems to be the only one that breaks using sanitycheck offline. `git grep` shows no other component using `FetchContent`. I found a few `ExternalProject` but they either don't download anything or are not part of the sanitycheck defaults.

vanwinkeljan commented:
@marc-hb I double checked with @nashif today and lvgl will be part of a second wave of migrating components out of tree

marc-hb commented:
Thx @vanwinkeljan . I realized the problem is not whether the lvgl module has its own git repo or not. From what I've seen so far, moving modules out of the main repo is a Good Thing But it never makes any module magically more modular or optional. Sorry for this west digression.

`FetchContent` is the real issue. `FetchContent` is just plain bad as it makes builds depend on an external resource and the network. There are (at least) two ways to fix this:

- Either lvgl is made truly optional and never ever included in any automated build, so no automated build ever fails unexpectedly unless it explicitly requested lvgl. Vote with "Thumbs up" emoji.
- Or  https://github.com/littlevgl/lvgl.git gets added as a new west module and true dependency and `FetchContent` is not needed anymore. Vote with "Hooray" emoji.
- Other, vote with "Thumbs down" emoji and elaborate.

vanwinkeljan commented:
@marc-hb `FetchContent` is just a stopgap solution until the lvgl dependency is migrated to west

marc-hb commented:
> lvgl will be part of a second wave of migrating components out of tree

This "lvgl" above sounded like the `zephyr/lib/gui/*` but you meant https://github.com/littlevgl/lvgl.git ?

vanwinkeljan commented:
@marc-hb yes with `lvgl` I did mean  https://github.com/littlevgl/lvgl.git; `zephyr/lib/gui/*` is merely glue logic
